### ISSUE TITLE
Eigen error with the C++ API and minibatches

### ISSUE THREAD

**Issue by FilippoC** [clab/dynet#640]
Hello,
I have the following error when training with minibatches :
`probabilistic-parser: /home/dist/corro/repos/eigen/Eigen/src/Core/Product.h:97: Eigen::Product<Lhs, Rhs, Option>::Product(const Lhs&, const Rhs&) [with _Lhs = Eigen::Map<Eigen::Matrix<float, -1, -1>, 0, Eigen::Stride<0, 0> >; _Rhs = Eigen::Map<Eigen::Matrix<float, -1, -1>, 0, Eigen::Stride<0, 0> >; int Option = 0; Eigen::Product<Lhs, Rhs, Option>::Lhs = Eigen::Map<Eigen::Matrix<float, -1, -1>, 0, Eigen::Stride<0, 0> >; Eigen::Product<Lhs, Rhs, Option>::Rhs = Eigen::Map<Eigen::Matrix<float, -1, -1>, 0, Eigen::Stride<0, 0> >]: Assertion `lhs.cols() == rhs.rows() && "invalid matrix product" && "if you wanted a coeff-wise or a dot product use the respective explicit functions"' failed. `

I am unable to use gdb on my cluster so I don't really know how I can debug this (the code is fairly huge).
Do you have any idea why this is not triggered by a dynet assertion ?

Thanks.

PS: sizes are weird (-1, -1), my guess is that it is the value used by Eigen for dynamic sizes ?

**Comment by neubig** [comment]
This is probably the result of a dimensionality error, and DyNet should indeed throw an error directly, but it's not catching the error properly. Could you please try to turn on immediate computation to locate the exact operation that is causing the problem? 
> http://dynet.readthedocs.io/en/latest/debugging.html#debugging-tools

Then if you can tell us the operation and the dimensions of the input, we should be able to both give you advice and fix the bug. Thanks!

**Comment by FilippoC** [comment]
Here is the stacktrace with immediate computation :

> 0x00002aaaac8db067 in __GI_raise (sig=sig@entry=6)                                                                                                                                                          [12/94]
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56                                                        
56      ../nptl/sysdeps/unix/sysv/linux/raise.c: Aucun fichier ou dossier de ce type.                    
#0  0x00002aaaac8db067 in __GI_raise (sig=sig@entry=6)                                                   
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56                                                        
#1  0x00002aaaac8dc448 in __GI_abort () at abort.c:89                                                    
#2  0x00002aaaac8d4266 in __assert_fail_base (      
    fmt=0x2aaaaca0cf18 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n",                                      
    assertion=assertion@entry=0x2aaaabd18f88 "lhs.cols() == rhs.rows() && \"invalid matrix product\" && \"if you wanted a coeff-wise or a dot product use the respective explicit functions\"",                    
    file=file@entry=0x2aaaabd18f50 "/home/dist/corro/repos/eigen/Eigen/src/Core/Product.h", line=line@entry=97,                                                                                                    
    function=function@entry=0x2aaaabd8e5c0 <Eigen::Product<Eigen::Map<Eigen::Matrix<float, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >, Eigen::Map<Eigen::Matrix<float, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0>
 >, 0>::Product(Eigen::Map<Eigen::Matrix<float, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > const&, Eigen::Map<Eigen::Matrix<float, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > const&)::__PRETTY_FUNCTION__> "Ei
gen::Product<Lhs, Rhs, Option>::Product(const Lhs&, const Rhs&) [with _Lhs = Eigen::Map<Eigen::Matrix<float, -1, -1>, 0, Eigen::Stride<0, 0> >; _Rhs = Eigen::Map<Eigen::Matrix<float, -1, -1>, 0, Eig"...)        
    at assert.c:92                                                                                       
#3  0x00002aaaac8d4312 in __GI___assert_fail (
    assertion=0x2aaaabd18f88 "lhs.cols() == rhs.rows() && \"invalid matrix product\" && \"if you wanted a coeff-wise or a dot product use the respective explicit functions\"",
    file=0x2aaaabd18f50 "/home/dist/corro/repos/eigen/Eigen/src/Core/Product.h", line=97,
    function=0x2aaaabd8e5c0 <Eigen::Product<Eigen::Map<Eigen::Matrix<float, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >, Eigen::Map<Eigen::Matrix<float, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >, 0>::Product
(Eigen::Map<Eigen::Matrix<float, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > const&, Eigen::Map<Eigen::Matrix<float, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > const&)::__PRETTY_FUNCTION__> "Eigen::Product<Lh
s, Rhs, Option>::Product(const Lhs&, const Rhs&) [with _Lhs = Eigen::Map<Eigen::Matrix<float, -1, -1>, 0, Eigen::Stride<0, 0> >; _Rhs = Eigen::Map<Eigen::Matrix<float, -1, -1>, 0, Eig"...) at assert.c:101
#4  0x00002aaaabbde335 in Product (rhs=..., lhs=..., this=0x7fffffffa950)
    at /home/dist/corro/repos/eigen/Eigen/src/Core/Product.h:95
#5  operator*<Eigen::Map<Eigen::Matrix<float, -1, -1>, 0, Eigen::Stride<0, 0> > > (other=..., this=<optimized out>)
    at /home/dist/corro/repos/eigen/Eigen/src/Core/GeneralProduct.h:394
#6  dynet::AffineTransform::forward_dev_impl<dynet::Device_CPU> (
    this=<optimized out>, dev=...,
    xs=std::vector of length 2, capacity 2 = {...}, fx=...)
    at /home/dist/corro/repos/dynet/dynet/nodes.cc:346
#7  0x00002aaaabb56dcf in dynet::AffineTransform::forward_impl (
    this=<optimized out>, xs=..., fx=...)
    at /home/dist/corro/repos/dynet/dynet/nodes.cc:450
#8  0x00002aaaababf53c in dynet::Node::forward (this=this@entry=0x176f6500,
    xs=std::vector of length 2, capacity 2 = {...}, fx=...)
    at /home/dist/corro/repos/dynet/dynet/dynet.cc:31
#9  0x00002aaaabaecd34 in dynet::BatchedExecutionEngine::incremental_forward_no_update (this=this@entry=0x1a223280, upto=..., autobatch_strategy=1)
    at /home/dist/corro/repos/dynet/dynet/exec.cc:742
#10 0x00002aaaabaf4ef9 in dynet::BatchedExecutionEngine::incremental_forward (
    this=0x1a223280, i=...) at /home/dist/corro/repos/dynet/dynet/exec.cc:778
#11 0x00002aaaabae665c in dynet::BatchedExecutionEngine::forward (
    this=0x1a223280, i=...) at /home/dist/corro/repos/dynet/dynet/exec.cc:328
#12 0x00002aaaabac1a6a in dynet::ComputationGraph::forward (
    this=this@entry=0x7fffffffb5e0, last=...)
    at /home/dist/corro/repos/dynet/dynet/dynet.cc:357


It seems that it comes from the following forward call :

> dynet::expr::Expression loss = dynet::expr::sum(batch_loss);
> cg.forward(loss);

I call the following after each CG creation :
> cg.set_immediate_compute(true);

Note : sorry, this is using an "old" dynet version. The code runs under commi 59b5bdd9a43fec30520951b61230c554a40058da

**Comment by FilippoC** [comment]
It seems that this dynet assertion fails to find the problem : https://github.com/clab/dynet/blob/59b5bdd9a43fec30520951b61230c554a40058da/dynet/nodes.cc#L344

**Comment by neubig** [comment]
Hmm, it should never get to that assertion...
Could you try adding the line `cg.print_graphviz()` before the first line that causes an error, and send the result? Also, are you using autobatching? If so, try adding `--dynet_autobatch_debug 1` as well.

**Comment by neubig** [comment]
I haven't heard back on this, so perhaps it's resolved?
